include/boost/math/special_functions/detail/simple_bitset.hpp

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 John Maddock and Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_DETAIL_SIMPLE_BITSET_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_DETAIL_SIMPLE_BITSET_HPP
+
+#include <memory>
+#include <cstring>
+#include <cstdint>
+#include <climits>
+
+namespace boost::math::detail::prime_sieve
+{
+template <typename I = std::uint64_t>
+class simple_bitset
```

Comment:
I ended up simplifying this code a lot locally as while msvc does a decent job of optimising this code, GCC does not.  And strangely both compilers generate faster code for `simple_bitset<unsigned char>` with the simplified version.  Here we go:

```
      template <class I>
      struct simple_bitset
      {
         std::unique_ptr<I[]> bits;
         std::size_t m_size;
         simple_bitset(std::size_t n) : bits(new I[n / (sizeof(I) * CHAR_BIT) + (n % (sizeof(I) * CHAR_BIT) ? 1 : 0)]), m_size(n)
         {
            std::memset(bits.get(), 0xff, n / CHAR_BIT + (n % CHAR_BIT ? 1 : 0));
         }
         // Initialize with specific pattern:
         simple_bitset(std::size_t n, I const* pattern, std::size_t len) : bits(new I[n / (sizeof(I) * CHAR_BIT) + (n % (sizeof(I) * CHAR_BIT) ? 1 : 0)]), m_size(n)
         {
            const std::size_t block_count = n / (sizeof(I) * CHAR_BIT) + (n % (sizeof(I) * CHAR_BIT) ? 1 : 0);
            if (block_count <= len)
            {
               std::memcpy(bits.get(), pattern, block_count * sizeof(I));
            }
            else
            {
               I* p = bits.get();
               std::memcpy(p, pattern, len * sizeof(I));
               I* base = p;
               p += len;
               while (len <= block_count - len)
               {
                  std::memcpy(p, base, len * sizeof(I));
                  p += len;
                  len *= 2;
               }
               if (block_count > len)
                  std::memcpy(p, base, (block_count - len) * sizeof(I));
            }
         }
         static constexpr std::size_t ln2(std::size_t n)
         {
            return n <= 1 ? 0 : 1 + ln2(n >> 1);
         }
         I* limbs()
         {
            return bits.get();
         }
         I test(std::size_t n)const
         {
            constexpr I mask = (sizeof(I) * CHAR_BIT) - 1;
            constexpr std::size_t shift = ln2(sizeof(I) * CHAR_BIT);
            return bits[n >> shift] & (I(1u) << (n & mask));
         }
         void clear(std::size_t n)
         {
            constexpr I mask = (sizeof(I) * CHAR_BIT) - 1;
            constexpr std::size_t shift = ln2(sizeof(I) * CHAR_BIT);
            bits[n >> shift] &= ~I(I(1u) << (n & mask));
         }
         void set(std::size_t n)
         {
            constexpr I mask = (sizeof(I) * CHAR_BIT) - 1;
            constexpr std::size_t shift = ln2(sizeof(I) * CHAR_BIT);
            //BOOST_ASSERT((n >> shift) < (m_size / (sizeof(I) * CHAR_BIT) + (m_size % (sizeof(I) * CHAR_BIT) ? 1 : 0)));
            bits[n >> shift] |= (I(1u) << (n & mask));
         }
         std::size_t size()const { return m_size; }
         void reset() { std::memset(bits.get(), 0xff, m_size / CHAR_BIT + (m_size % CHAR_BIT ? 1 : 0)); }
         void clear_all() { std::memset(bits.get(), 0, m_size / CHAR_BIT + (m_size % CHAR_BIT ? 1 : 0)); }
      };
```

Reply:
The updated bitset has been added in commit d649f65. For initializing with a specific pattern what do you think about using the [gaps in the MOD 30 wheel](https://en.wikipedia.org/wiki/Wheel_factorization#/media/File:Wheel_factorization-n=30.svg) as the specific pattern to set? This would remove setting 22 of every 30 values and is a very short pattern. The only difficulty would be finding where to start based of each threads assigned lower and upper bound. The `IntervalSieve` class is reused so it would only be a concern on the first pass. `simple_bitset` should then satisfy the requirements for packed data structure and the initialization optimization.